modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Services/ServicesClient.cs

```diff
@@ -202,7 +215,123 @@ public IServicesClient WithServerKeepBinary()
+                    return;
+
+                var socket = _srvc._ignite.Socket;
+
+                var topVer = socket.GetTopologyVersion();
+
+                var log = _srvc._ignite.GetConfiguration().Logger;
+
+                var groupNodes = _srvc._clusterGroup?.GetNodes();
+                
+                _nodes = socket.DoOutInOpAsync(ClientOp.ServiceGetTopology,
+                    ctx => ctx.Writer.WriteString(_srvcName),
+                    ctx =>
+                    {
+                        var cnt = ctx.Reader.ReadInt();
+
+                        IList<Guid> res = new List<Guid>(cnt);
+
+                        for (var i = 0; i < cnt; ++i)
+                            res.Add(BinaryUtils.ReadGuid(ctx.Reader.Stream));
+
+                        Interlocked.Exchange(ref _lastUpdateRequestTime, DateTime.Now.Ticks);
+                        Interlocked.Exchange(ref _lastAffTop, topVer);
+                        Interlocked.Exchange(ref _updateInProgress, 0);
+
+                        var filteredTopology = FilterTopology(res, groupNodes?.Select(n => n.Id).ToList());
+
+                        log.Debug("Topology of service '" + _srvcName + "' has been updated. The " +
+                                  "service instance nodes: " + string.Join(", ", res.Select(gid=>gid.ToString())) +
+                                  ". Effective topology with the cluster group is: " + 
+                                  string.Join(", ", filteredTopology.Select(gid=>gid.ToString())) + '.');
+
+                        return filteredTopology;
+                    }, (status, err) =>
+                    {
+                        Interlocked.Exchange(ref _updateInProgress, 0);
+
+                        log.Error("Failed to update topology of the service '" + _srvcName + "'.", err);
+
+                        return _nodes;
+                    }).Result;
+            }
+
+            /// <summary>
+            /// Filters service topology regarding to the cluster group.
+            /// </summary>
+            private static IList<Guid> FilterTopology(IList<Guid> serviceTopology, IList<Guid> clusterGroup)
+            {
+                return clusterGroup == null ? serviceTopology : serviceTopology.Intersect(clusterGroup).ToList();
+            }
+
+            /// <summary>
+            /// Provides last known service topology and asynchronously updates it if required.
+            /// </summary>
+            internal IList<Guid> GetAndUpdate()
+            {
+                long lastKnownAff = Interlocked.Read(ref _lastAffTop);
+                long curAff = _srvc._ignite.Socket.GetTopologyVersion();
+                long lastUpdateTime = Interlocked.Read(ref _lastUpdateRequestTime);
+                
+                if(lastKnownAff == 0 || curAff > lastKnownAff || DateTime.Now.Ticks - lastUpdateTime >= SrvTopUpdatePeriod)
+                    UpdateTopologyAsync();
```

Comment:
What's the idea - do we want to update in background and return current nodes meanwhile?

Reply:
Yes. Just like in Java Thin. The same.